src/expr/e_parser.ml

```diff
@@ -594,6 +672,175 @@ and boundeds b = lazy begin
   end
 end
 
+(* comma-separated listed of declarations within function constructor *)
+and func_boundeds b = lazy begin
+    (* comma-separated list of declarations, examples:
+
+    m \in S
+    a, b \in R
+    m \in S,  a, b \in R
+    m \in S,  a, b \in R,  <<x, y>> \in A \X B
+    *)
+    sep1 (punct ",") (choice [
+        (* bounded constants, examples:
+
+        m \in S
+        a, b \in R
+        *)
+        (sep1 (punct ",") hint <*> (infix "\\in" >*> use (expr b)))
+            <$> begin
+                fun (vs, dom) ->
+                    let bounds =
+                        let hd = (List.hd vs, Constant, Domain dom) in
+                        let tl = List.map
+                            (fun v -> (v, Constant, Ditto)) (List.tl vs) in
+                        hd :: tl in
+                    let letin = [] in
+                    (bounds, letin)
+            end;
+        (* bounded tuples of constants, examples:
+
+        <<x, y>> \in S
+        <<a, b, c >> \in A \X B \X C
+
+        A function constructor is represented with `Fcn`,
+        which takes `bounds` as first argument.
+        `bounds` represents a list of bound identifiers (constants here).
+
+        So the tuples need to be converted to individual identifier bounds.
+        This is done by introducing intermediate definitions in a `LET...IN`.
+        Each bounded tuple (like `<<x, y>>` above) is replaced by a fresh
+        identifier of the form:
+
+            fcnbnd#concatenated_names
+
+        where "concatenated_names" results from concatenating the identifiers
+        that occur within the tuple. For example, `<<x, y>>` is replaced by
+
+            fcnbnd#xy
+
+        The fresh identifier is used inside the `LET...IN` for defining each
+        of the identifiers that occurred within the tuple. For example,
+        `[<<x, y>> \in S |-> ...]` becomes:
+
+            [fcnbnd#xy \in S:
+                LET
+                    x == fcnbnd#xy[1]
+                    y == fcnbnd#xy[2]
+                IN
+                    ...]
+
+        The hashmark is used within the identifier fcnbnd#... to ensure that
+        the fresh identifier is different from all other identifiers in the
+        current context, without the need to inspect the context (which is
+        not available while parsing). The syntax of TLA+ ensures this,
+        because no identifier in TLA+ source can contain a hashmark.
+
+        In each function constructor, the concatenation of identifiers
+        (like "xy" above) is unique, because the TLA+ syntax ensures that
+        each identifier is unique within its context. Therefore, each bounded
```

Comment:
Thank you for finding this error. Indeed, this would result in two identifiers called `fcnbnd#xyz`. This error is demonstrated in the output:

```
ASSUME g ==
         [fcnbnd#ab \in {<<1>>},
            fcnbnd#ab_1 \in {2} \X {3} |->
            LET ab == fcnbnd#ab_1[1]
                a == fcnbnd#ab_1[1]
                b == fcnbnd#ab_1[2]
            IN  ab]
PROVE  g[<<1>>, <<2, 3>>] = 1
```

that is generated for `g == [<<ab>> \in {<<1>>}, <<a, b>> \in {2} \X {3} |-> ab]`, from a new commented test that would catch this bug if the backends supported the expressions involved in the test.

I believe that I now have corrected the error, by using only the first identifier from the tuple declaration, as suggested.